<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Model Diagnostics for RNAseq Data — evaluateDist • powsimR</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Model Diagnostics for RNAseq Data — evaluateDist" />

<meta property="og:description" content="With this function, the user can determine goodness of fit for each gene." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">powsimR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/powsimR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/bvieth/powsimR">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Model Diagnostics for RNAseq Data</h1>
    <small class="dont-index">Source: <a href='https://github.com/bvieth/powsimR/blob/master/R/Evaluate.R'><code>R/Evaluate.R</code></a></small>
    <div class="hidden name"><code>evaluateDist.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>With this function, the user can determine goodness of fit for each gene.</p>
    
    </div>

    <pre class="usage"><span class='fu'>evaluateDist</span>(<span class='no'>countData</span>, <span class='kw'>batchData</span> <span class='kw'>=</span><span class='kw'>NULL</span>,
<span class='kw'>spikeData</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>spikeInfo</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
<span class='kw'>Lengths</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>MeanFragLengths</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
<span class='no'>RNAseq</span>, <span class='no'>Normalisation</span>,
<span class='kw'>frac.genes</span><span class='kw'>=</span><span class='fl'>1</span>, <span class='kw'>min.meancount</span> <span class='kw'>=</span> <span class='fl'>0.1</span>,
<span class='kw'>max.dropout</span><span class='kw'>=</span><span class='fl'>0.7</span>, <span class='kw'>min.libsize</span><span class='kw'>=</span><span class='fl'>1000</span>,
<span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>countData</th>
      <td><p>is a count matrix (row=gene, column=sample).
Please provide the measurements of one group only, e.g. the control group.</p></td>
    </tr>
    <tr>
      <th>batchData</th>
      <td><p>is a <code>data.frame</code> for batch annotation.
Rows correspond to samples. The first column should contain the batches, e.g. 'a', 'b', 'c', etc..
This is only used for the normalisation.</p></td>
    </tr>
    <tr>
      <th>spikeData</th>
      <td><p>is a count <code>matrix</code>.
Rows correspond to spike-ins, columns to samples.
The order of columns should be the same as in the <code>countData</code>.</p></td>
    </tr>
    <tr>
      <th>spikeInfo</th>
      <td><p>is a molecule count <code>matrix</code> of spike-ins.
Rows correspond to spike-ins. The order of rows should be the same as in the <code>spikeData</code>.
The column names should be 'SpikeID' and 'SpikeInput' for molecule counts of spike-ins.</p></td>
    </tr>
    <tr>
      <th>Lengths</th>
      <td><p>is a numeric vector of transcript lengths with the same length and order as the rows in countData.
This variable is only used for internal TPM calculations if Census normalization is specified.</p></td>
    </tr>
    <tr>
      <th>MeanFragLengths</th>
      <td><p>is a numeric vector of mean fragment lengths with the same length as columns in countData.
This variable is only used for internal TPM calculations if Census normalization is specified.</p></td>
    </tr>
    <tr>
      <th>RNAseq</th>
      <td><p>is a character value: "bulk" or "singlecell".</p></td>
    </tr>
    <tr>
      <th>Normalisation</th>
      <td><p>is a character value: 'TMM', 'MR', 'PosCounts', 'UQ', 'scran', 'Linnorm',
'SCnorm', 'RUV', 'Census', 'depth', 'none'.
For more information, please consult the details section of <code><a href='estimateParam.html'>estimateParam</a></code>.</p></td>
    </tr>
    <tr>
      <th>frac.genes</th>
      <td><p>The fraction of genes to calculate goodness of fit statistics, default is 1, i.e. for all genes.</p></td>
    </tr>
    <tr>
      <th>min.meancount</th>
      <td><p>The minimum mean normalised count per gene if fraction of genes are defined. Default is <code>0.1</code>.</p></td>
    </tr>
    <tr>
      <th>max.dropout</th>
      <td><p>The maximal percentage of zero expression per gene. Genes with more than <code>max.dropout</code> will be remove. Default is <code>0.7</code>, i.e. genes with more than 70% dropouts.</p></td>
    </tr>
    <tr>
      <th>min.libsize</th>
      <td><p>The minimum raw read counts per sample, default is <code>1000</code>.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>Logical value to indicate whether to show progress report of simulations.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>List object with the results of goodness of fit and estimated parameters:</p>
<dt>edgeR</dt><dd><p>Goodness-of-fit statistic, degrees of freedom and associated p-value using the deviance and residual degrees of freedom from <code><a href='https://www.rdocumentation.org/packages/edgeR/topics/glmfit'>glmFit</a></code>. Furthermore, the AIC of the edgeR model fit using the residuals of <code><a href='https://www.rdocumentation.org/packages/edgeR/topics/zscoreNBinom'>zscoreNBinom</a></code>.</p></dd>
<dt>GOF</dt><dd><p>The fitting results per distribution, including loglikelihood, goodness-of-fit statistics, AIC and predicted number of zeroes. The following distributions were considered: Poisson, negative binomial, zero-inflated poisson and negative binomial following the 'standard' (i.e. <code><a href='https://www.rdocumentation.org/packages/stats/topics/glm'>glm</a></code>, <code><a href='https://www.rdocumentation.org/packages/MASS/topics/glm.nb'>glm.nb</a></code> and <code><a href='https://www.rdocumentation.org/packages/pscl/topics/zeroinfl'>zeroinfl</a></code> implementation) and fitdist approach (see <code><a href='https://www.rdocumentation.org/packages/fitdistrplus/topics/fitdist'>fitdist</a></code>) and Beta-Poisson following Marioni or Hemberg parameterisation. Furthermore, model fit comparison by LRT for nested and Vuong Test for non-nested models.</p></dd>
<dt>Estimates</dt><dd><p>The estimated parameters of distribution fitting.</p></dd>
<dt>ObservedZeros</dt><dd><p>The number of zeroes and dropout rate per gene.</p></dd>

    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'># NOT RUN {</span>
<span class='co'>## using example data set</span>
<span class='fu'><a href='https://www.rdocumentation.org/packages/utils/topics/data'>data</a></span>(<span class='no'>kolodziejczk_cnts</span>)
<span class='no'>evaldist</span> <span class='kw'>&lt;-</span> <span class='fu'>evaluateDist</span>(<span class='kw'>countData</span> <span class='kw'>=</span> <span class='no'>kolodziejczk_cnts</span>,
<span class='kw'>RNAseq</span> <span class='kw'>=</span> <span class='st'>"singlecell"</span>, <span class='kw'>Normalisation</span><span class='kw'>=</span><span class='st'>"scran"</span>,
<span class='kw'>frac.genes</span><span class='kw'>=</span><span class='fl'>1</span>, <span class='kw'>min.meancount</span> <span class='kw'>=</span> <span class='fl'>0.1</span>,
<span class='kw'>max.dropout</span><span class='kw'>=</span><span class='fl'>0.7</span>, <span class='kw'>min.libsize</span><span class='kw'>=</span><span class='fl'>1000</span>,
<span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)
<span class='fu'><a href='plotEvalDist.html'>plotEvalDist</a></span>(<span class='no'>evaldist</span>, <span class='kw'>annot</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)
<span class='co'># }</span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Beate Vieth</p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Beate Vieth.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.2.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

