<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>In Silico Simulation Parameters — insilicoNBParam • powsimR</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="In Silico Simulation Parameters — insilicoNBParam" />

<meta property="og:description" content="With this function, the user can define the parameters of the negative binomial distribution needed for the simulations. The mean, dispersion and dropout need to be specified for bulk RNAseq experiments. For single cell RNAseq experiments, the mean and dispersion should be chosen so that a high number of dropouts are generated (see details)." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">powsimR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/powsimR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/bvieth/powsimR">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>In Silico Simulation Parameters</h1>
    <small class="dont-index">Source: <a href='https://github.com/bvieth/powsimR/blob/master/R/Param.R'><code>R/Param.R</code></a></small>
    <div class="hidden name"><code>insilicoNBParam.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>With this function, the user can define the parameters of the negative binomial distribution needed for the simulations. The mean, dispersion and dropout need to be specified for bulk RNAseq experiments. For single cell RNAseq experiments, the mean and dispersion should be chosen so that a high number of dropouts are generated (see details).</p>
    
    </div>

    <pre class="usage"><span class='fu'>insilicoNBParam</span>(<span class='no'>means</span>, <span class='no'>dispersion</span>,
<span class='kw'>dropout</span><span class='kw'>=</span><span class='kw'>NULL</span>, <span class='kw'>sf</span><span class='kw'>=</span><span class='kw'>NULL</span>, <span class='kw'>RNAseq</span><span class='kw'>=</span><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"bulk"</span>, <span class='st'>"singlecell"</span>))</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>means</th>
      <td><p>mean parameter of the NB read count distribution defined by a function sampling from a random distribution, e.g. function(x) rgamma(n=x, shape=2, rate=2).</p></td>
    </tr>
    <tr>
      <th>dispersion</th>
      <td><p>dispersion parameter of the NB read count distribution. This can be:
(1) a constant, e.g. 0.2
(2) a function relating to the mean expression. The parametric fit employed in DESeq2 is recommended to get an estimate, e.g. function(x) 3 + 150/x.</p></td>
    </tr>
    <tr>
      <th>dropout</th>
      <td><p>The probability of droput per gene. This can be:
(1) a constant
(2) a function sampling from a random distribution, e.g. function(x) runif(x, min=0, max=0.25).
The default is <code>NULL</code>, i.e. no dropout. Note that the dropout is not considered for single cell simulations. This should be inherently given by low mean expression coupled with high dispersion.</p></td>
    </tr>
    <tr>
      <th>sf</th>
      <td><p>The size factor:
(1) a vector
(2) a function sampling from a random distribution, e.g. function(x) rnorm(x, mean=1, sd=0.2).
The default is <code>NULL</code>, i.e. equal size factor of 1.</p></td>
    </tr>
    <tr>
      <th>RNAseq</th>
      <td><p>is a character value: "bulk" or "singlecell".</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>List with the following vectors:</p>
<dt>means</dt><dd><p>Mean log2 read counts per gene.</p></dd>
<dt>dispersion</dt><dd><p>Log2 dispersion per gene.</p></dd>
<dt>p0</dt><dd><p>Probability that the count will be zero per gene.</p></dd>
<dt>sf</dt><dd><p>Size factor per sample.</p></dd>
<dt>estFramework</dt><dd><p>"in silico" by default.</p></dd>

    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'># NOT RUN {</span>
<span class='co'>## Bulk RNA-seq experiment in silico parameters:</span>
<span class='no'>insilico.bulk</span> <span class='kw'>&lt;-</span> <span class='fu'>insilicoNBParam</span>(<span class='kw'>means</span><span class='kw'>=</span><span class='kw'>function</span>(<span class='no'>x</span>) <span class='fl'>2</span>^<span class='fu'><a href='https://www.rdocumentation.org/packages/stats/topics/Uniform'>runif</a></span>(<span class='no'>x</span>, <span class='fl'>9</span>, <span class='fl'>12</span>),
<span class='kw'>dispersion</span><span class='kw'>=</span><span class='fl'>0.2</span>,
<span class='kw'>dropout</span><span class='kw'>=</span><span class='kw'>function</span>(<span class='no'>x</span>) <span class='fu'><a href='https://www.rdocumentation.org/packages/stats/topics/Uniform'>runif</a></span>(<span class='no'>x</span>, <span class='kw'>min</span> <span class='kw'>=</span> <span class='fl'>0</span>, <span class='kw'>max</span> <span class='kw'>=</span> <span class='fl'>0.25</span>),
<span class='kw'>sf</span><span class='kw'>=</span><span class='kw'>function</span>(<span class='no'>x</span>) <span class='kw pkg'>truncnorm</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/truncnorm/topics/dtruncnorm'>rtruncnorm</a></span>(<span class='no'>x</span>, <span class='kw'>a</span> <span class='kw'>=</span> <span class='fl'>0.6</span>, <span class='kw'>b</span> <span class='kw'>=</span> <span class='fl'>1.4</span>, <span class='kw'>mean</span><span class='kw'>=</span><span class='fl'>1</span>, <span class='kw'>sd</span><span class='kw'>=</span><span class='fl'>0.2</span>),
<span class='kw'>RNAseq</span><span class='kw'>=</span><span class='st'>'bulk'</span>)
<span class='co'>## Single cell RNA-seq experiment in silico parameters:</span>
<span class='no'>insilico.singlecell</span> <span class='kw'>&lt;-</span> <span class='fu'>insilicoNBParam</span>(<span class='kw'>means</span><span class='kw'>=</span><span class='kw'>function</span>(<span class='no'>x</span>) <span class='fu'><a href='https://www.rdocumentation.org/packages/stats/topics/GammaDist'>rgamma</a></span>(<span class='no'>x</span>, <span class='fl'>4</span>, <span class='fl'>2</span>),
<span class='kw'>dispersion</span><span class='kw'>=</span><span class='kw'>function</span>(<span class='no'>x</span>) <span class='fl'>2</span> + <span class='fl'>150</span>/<span class='no'>x</span>,
<span class='kw'>dropout</span><span class='kw'>=</span><span class='kw'>NULL</span>,
<span class='kw'>sf</span><span class='kw'>=</span><span class='kw'>function</span>(<span class='no'>x</span>) <span class='fl'>2</span>^<span class='fu'><a href='https://www.rdocumentation.org/packages/stats/topics/Normal'>rnorm</a></span>(<span class='kw'>n</span><span class='kw'>=</span><span class='no'>x</span>, <span class='kw'>mean</span><span class='kw'>=</span><span class='fl'>0</span>, <span class='kw'>sd</span><span class='kw'>=</span><span class='fl'>0.5</span>),
<span class='kw'>RNAseq</span><span class='kw'>=</span><span class='st'>'singlecell'</span>)
<span class='co'># }</span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Beate</p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Beate Vieth.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.2.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

