<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Simulate Differential Expression — simulateDE • powsimR</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Simulate Differential Expression — simulateDE" />

<meta property="og:description" content="simulateDE is the main function to simulate differential expression for RNA-seq experiments.
The simulation parameters are specified with SimSetup.
The user needs to specify furthermore
the number of samples per group, preprocessing, normalisation and differential testing method.
There is also the option to consider spike-ins. 
The return object contains DE test results from all simulations as well as descriptive statistics.
The error matrix calculations will be conducted with evaluateDE." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">powsimR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/powsimR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/bvieth/powsimR">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Simulate Differential Expression</h1>
    <small class="dont-index">Source: <a href='https://github.com/bvieth/powsimR/blob/master/R/Simulate.R'><code>R/Simulate.R</code></a></small>
    <div class="hidden name"><code>simulateDE.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>simulateDE is the main function to simulate differential expression for RNA-seq experiments.
The simulation parameters are specified with <code><a href='SimSetup.html'>SimSetup</a></code>.
The user needs to specify furthermore
the number of samples per group, preprocessing, normalisation and differential testing method.
There is also the option to consider spike-ins. <br />
The return object contains DE test results from all simulations as well as descriptive statistics.
The error matrix calculations will be conducted with <code><a href='evaluateDE.html'>evaluateDE</a></code>.<br /></p>
    
    </div>

    <pre class="usage"><span class='fu'>simulateDE</span>(<span class='kw'>n1</span><span class='kw'>=</span><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='fl'>20</span>,<span class='fl'>50</span>,<span class='fl'>100</span>), <span class='kw'>n2</span><span class='kw'>=</span><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='fl'>30</span>,<span class='fl'>60</span>,<span class='fl'>120</span>),
<span class='no'>sim.settings</span>,
<span class='no'>DEmethod</span>,
<span class='no'>Normalisation</span>,
<span class='kw'>Label</span> <span class='kw'>=</span> <span class='st'>"none"</span>,
<span class='kw'>Prefilter</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
<span class='kw'>Impute</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
<span class='kw'>DEFilter</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
<span class='kw'>spikeIns</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
<span class='kw'>NCores</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
<span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>n1, n2</th>
      <td><p>Integer vectors specifying the number of biological replicates in each group.
Default values are n1=c(20,50,100) and n2=c(30,60,120).</p></td>
    </tr>
    <tr>
      <th>sim.settings</th>
      <td><p>This object specifies the simulation setup. This must be the return object from <code><a href='SimSetup.html'>SimSetup</a></code>.</p></td>
    </tr>
    <tr>
      <th>DEmethod</th>
      <td><p>A character vector specifying the DE detection method to be used.
Please consult the Details section for available options.</p></td>
    </tr>
    <tr>
      <th>Normalisation</th>
      <td><p>Normalisation method to use.
Please consult the Details section for available options.</p></td>
    </tr>
    <tr>
      <th>Label</th>
      <td><p>A character vector to define whether information about group labels should be used for normalisation.
This is only implemented for scran and SCnorm. Possible options include the default <code>"none"</code> which means that no sample group information is considered for normalisation; <code>"known"</code> means that the simulated group labels are used and <code>"clustering"</code> which applies an unsupervised hierarchical clustering to determine the group labels (for details, see <code><a href='https://www.rdocumentation.org/packages/scran/topics/quickCluster'>quickCluster</a></code>).</p></td>
    </tr>
    <tr>
      <th>Prefilter</th>
      <td><p>A character vector specifying the gene expression filtering method
to be used prior to normalisation (and possibly imputation).
Default is <code>NULL</code>, i.e. no filtering.
Please consult the Details section for available options.</p></td>
    </tr>
    <tr>
      <th>Impute</th>
      <td><p>A character vector specifying the gene expression imputation method
to be used prior to normalisation.
Default is <code>NULL</code>, i.e. no imputation.
Please consult the Details section for available options.</p></td>
    </tr>
    <tr>
      <th>DEFilter</th>
      <td><p>A logical vector indicating whether to run DE testing on filtered and/or imputed count data.
Default is <code>FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>spikeIns</th>
      <td><p>Logical vector to indicate whether to simulate spike-ins.
Default is <code>FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>NCores</th>
      <td><p>integer positive number of cores for parallel processing.
Default is <code>NULL</code>, i.e. 1 core.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>Logical vector to indicate whether to show progress report of simulations.
Default is <code>TRUE</code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list with the following fields:</p>
<dt>pvalue, fdr</dt><dd><p>3D array (ngenes * N * nsims) for p-values and FDR from each simulation.
Note that FDR values will be empty and the calculation will be done by <code><a href='evaluateDE.html'>evaluateDE</a></code> whenever applicable.</p></dd>
<dt>mu,disp,dropout</dt><dd><p>3D (ngenes * N * nsims) array for mean, dispersion and dropout of library size factor normalized read counts.</p></dd>
<dt>true.mu,true.disp,true.dropout</dt><dd><p>3D (ngenes * N * nsims) array for true mean, dispersion and dropout of simulated read counts.</p></dd>
<dt>true.depth</dt><dd><p>True simulated sequencing depth per sample.</p></dd>
<dt>est.sf</dt><dd><p>Global library size factor estimates per sample.</p></dd>
<dt>est.gsf</dt><dd><p>3D array (ngenes * N * nsims) for size factor estimates. These are gene- and sample-wise estimates and only for SCnorm and Linnorm normalisation.</p></dd>
<dt>elfc,rlfc</dt><dd><p>3D array (ngenes * N * nsims) for log2 fold changes (LFC):
elfc is for the DE tool estimated LFC; rlfc is for the LFC estimated from the normalised read counts.</p></dd>
<dt>sim.settings</dt><dd><p>The input sim.settings to which the specifications of <code>simulateDE</code> is added.</p></dd>
<dt>time.taken</dt><dd><p>The time taken for each simulation, given for preprocessing, normalisation, differential expression testing and moment estimation.</p></dd>

    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Here you can find detailed information about preprocessing, imputation, normalisation and differential testing choices.</p>
    
    <h2 class="hasAnchor" id="prefiltering-prior-to-imputation-normalisation"><a class="anchor" href="#prefiltering-prior-to-imputation-normalisation"></a>Prefiltering prior to imputation/normalisation</h2>

    
    <dl class='dl-horizontal'>
<dt>CountFilter</dt><dd><p>removes genes that have a mean expression below 0.2.</p></dd>
<dt>FreqFilter</dt><dd><p>removes genes that have more than 80 percent dropouts.</p></dd>
</dl>
    
    <h2 class="hasAnchor" id="imputation-prior-to-normalisation"><a class="anchor" href="#imputation-prior-to-normalisation"></a>Imputation prior to normalisation</h2>

    
    <dl class='dl-horizontal'>
<dt>scImpute</dt><dd><p>employs scImpute method of imputing dropouts. Imputation is only carried out for genes with more than 50 percent dropout. Please consider multiple cores to speed up computation.</p></dd>
<dt>DrImpute</dt><dd><p>employs DrImpute method of imputing dropouts as implemented
in <code><a href='https://www.rdocumentation.org/packages/DrImpute/topics/DrImpute'>DrImpute</a></code>.</p></dd>
<dt>SAVER</dt><dd><p>employs SAVER method of imputing dropouts as implemented
in <code><a href='https://www.rdocumentation.org/packages/SAVER/topics/saver'>saver</a></code>. Imputation is only carried out for genes with more than 50 percent dropout.</p></dd>
<dt>Seurat</dt><dd><p>employs Seurat method of imputing dropouts as implemented
in <code><a href='https://www.rdocumentation.org/packages/Seurat/topics/AddImputedScore'>AddImputedScore</a></code> using variable genes identified with <code><a href='https://www.rdocumentation.org/packages/Seurat/topics/FindVariableGenes'>FindVariableGenes</a></code>. Imputation is only carried out for genes with more than 50 percent dropout.</p></dd>
<dt>scone</dt><dd><p>employs scone method of imputing dropouts as implemented
in <code><a href='https://www.rdocumentation.org/packages/scone/topics/scone'>scone</a></code> using estimated dropout probabilities of <code><a href='https://www.rdocumentation.org/packages/scone/topics/estimate_ziber'>estimate_ziber</a></code>.</p></dd>
</dl>
    
    <h2 class="hasAnchor" id="normalisation-applied-to-imputed-read-count-matrix"><a class="anchor" href="#normalisation-applied-to-imputed-read-count-matrix"></a>Normalisation applied to (imputed) read count matrix</h2>

    
    <dl class='dl-horizontal'>
<dt>TMM, UQ</dt><dd><p>employ the edgeR style normalization of weighted trimmed mean of M-values and upperquartile
as implemented in <code><a href='https://www.rdocumentation.org/packages/edgeR/topics/calcNormFactors'>calcNormFactors</a></code>, respectively.</p></dd>
<dt>MR, PosCounts</dt><dd><p>employ the DESeq2 style normalization of median ratio method and a modified geometric mean method
as implemented in <code><a href='https://www.rdocumentation.org/packages/DESeq2/topics/estimateSizeFactors'>estimateSizeFactors</a></code>, respectively. Spike-ins can also be supplied for both methods via <code>spikeData</code>.</p></dd>
<dt>scran, SCnorm</dt><dd><p>apply the deconvolution and quantile regression normalization methods developed for sparse RNA-seq data
as implemented in <code><a href='https://www.rdocumentation.org/packages/scran/topics/computeSumFactors'>computeSumFactors</a></code> and <code><a href='https://www.rdocumentation.org/packages/SCnorm/topics/SCnorm'>SCnorm</a></code>, respectively. Spike-ins can also be supplied for both methods via <code>spikeData</code>. Note, however that this means for scran that the normalisation as implemented in <code><a href='https://www.rdocumentation.org/packages/scran/topics/computeSpikeFactors'>computeSpikeFactors</a></code> is also applied to genes (<code>general.use=TRUE</code>). Please consider multiple cores to speed up computation for SCnorm.</p></dd>
<dt>Linnorm</dt><dd><p>apply the normalization method for sparse RNA-seq data
as implemented in <code><a href='https://www.rdocumentation.org/packages/Linnorm/topics/Linnorm.Norm'>Linnorm.Norm</a></code>.
For <code>Linnorm</code>, the user can also supply <code>spikeData</code>.</p></dd>
<dt>RUV</dt><dd><p>removes unwanted variation. There are two approaches implemented:
(1) utilizing negative control genes, i.e. spike-ins stored in <code>spikeData</code> (<code><a href='https://www.rdocumentation.org/packages/RUVSeq/topics/RUVg'>RUVg</a></code>).
(2) utilizing replicate samples, i.e. samples for which the covariates of interest are considered constant.
This annotation is stored in <code>batchData</code> (<code><a href='https://www.rdocumentation.org/packages/RUVSeq/topics/RUVs'>RUVs</a></code>).</p></dd>
<dt>Census</dt><dd><p>converts relative measures of TPM/FPKM values into mRNAs per cell (RPC) without the need of spike-in standards.
Census at least needs <code>Lengths</code> for single-end data and preferably <code>MeanFragLengths</code> for paired-end data.
Do not use this algorithm for UMI data!</p></dd>
<dt>depth</dt><dd><p>Sequencing depth normalisation.</p></dd>
</dl>
    
    <h2 class="hasAnchor" id="differential-testing-using-raw-read-count-matrix"><a class="anchor" href="#differential-testing-using-raw-read-count-matrix"></a>Differential testing using raw read count matrix</h2>

    
    <dl class='dl-horizontal'>
<dt>T-Test</dt><dd><p>A T-Test per gene is applied using log2 transformed and normalized expression values (i.e. CPM or TPM).</p></dd>
<dt>limma-trend, limma-voom</dt><dd><p>apply differential testing as implemented in <code><a href='https://www.rdocumentation.org/packages/limma/topics/lmFit'>lmFit</a></code>
followed by <code><a href='https://www.rdocumentation.org/packages/limma/topics/ebayes'>eBayes</a></code> on counts transformed by <code><a href='https://www.rdocumentation.org/packages/limma/topics/voom'>voom</a></code> or by applying mean-variance trend on log2 CPM values in <code><a href='https://www.rdocumentation.org/packages/limma/topics/ebayes'>eBayes</a></code>.</p></dd>
<dt>edgeR-LRT, edgeR-QL</dt><dd><p>apply differential testing as implemented in <code><a href='https://www.rdocumentation.org/packages/edgeR/topics/glmfit'>glmFit</a></code>, <code><a href='https://www.rdocumentation.org/packages/edgeR/topics/glmfit'>glmLRT</a></code> and<code><a href='https://www.rdocumentation.org/packages/edgeR/topics/glmQLFTest'>glmQLFit</a></code>, <code><a href='https://www.rdocumentation.org/packages/edgeR/topics/glmQLFTest'>glmQLFTest</a></code>, respectively.</p></dd>
<dt>DESeq2</dt><dd><p>applies differential testing as implemented in <code><a href='https://www.rdocumentation.org/packages/DESeq2/topics/DESeq'>DESeq</a></code>.</p></dd>
<dt>ROTS</dt><dd><p>applies differential testing as implemented in <code><a href='https://www.rdocumentation.org/packages/ROTS/topics/ROTS'>ROTS</a></code> with 100 permutations on transformed counts (<code><a href='https://www.rdocumentation.org/packages/limma/topics/voom'>voom</a></code>).</p></dd>
<dt>baySeq</dt><dd><p>applies differential testing as implemented in <code><a href='https://www.rdocumentation.org/packages/baySeq/topics/getLikelihoods'>getLikelihoods</a></code> based on negative binomial prior estimates (<code><a href='https://www.rdocumentation.org/packages/baySeq/topics/getPriors'>getPriors.NB</a></code>).</p></dd>
<dt>NOISeq</dt><dd><p>applies differential testing as implemented in <code><a href='https://www.rdocumentation.org/packages/NOISeq/topics/noiseqbio'>noiseqbio</a></code> based on CPM values.</p></dd>
<dt>EBSeq</dt><dd><p>applies differential testing as implemented in <code><a href='https://www.rdocumentation.org/packages/EBSeq/topics/EBTest'>EBTest</a></code>.</p></dd>
<dt>MAST</dt><dd><p>applies differential testing as implemented in <code><a href='https://www.rdocumentation.org/packages/MAST/topics/zlm'>zlm</a></code> for zero-inflated model fitting followed by <code><a href='https://www.rdocumentation.org/packages/MAST/topics/lrTest'>lrTest</a></code> on log2 CPM values.</p></dd>
<dt>scde</dt><dd><p>applies differential testing as implemented in <code><a href='https://www.rdocumentation.org/packages/scde/topics/scde.expression.difference'>scde.expression.difference</a></code>.</p></dd>
<dt>BPSC</dt><dd><p>applies differential testing as implemented in <code><a href='https://www.rdocumentation.org/packages/BPSC/topics/BPglm'>BPglm</a></code> on CPM values.</p></dd>
<dt>scDD</dt><dd><p>applies differential testing as implemented in <code><a href='https://www.rdocumentation.org/packages/scDD/topics/scDD'>scDD</a></code> on CPM values.</p></dd>
<dt>DECENT</dt><dd><p>applies differential testing as implemented in <code><a href='https://www.rdocumentation.org/packages/DECENT/topics/decent'>decent</a></code>.</p></dd>
<dt>edgeR-zingeR, DESeq2-zingeR</dt><dd><p>In a first step, the posterior probabilities of the zero-inflated negative binomial component are estimated (see <code><a href='https://www.rdocumentation.org/packages/zingeR/topics/zeroWeightsLS'>zeroWeightsLS</a></code>) and used to define a weight matrix for dispersion estimation in <code><a href='https://www.rdocumentation.org/packages/edgeR/topics/estimateDisp'>estimateDisp</a></code>. For the edgeR approach, the generalized model as implemented in <code><a href='https://www.rdocumentation.org/packages/edgeR/topics/glmfit'>glmFit</a></code> is fitted. This is followed by an adapted LRT for differential testing to account for the weighting (see <code><a href='https://www.rdocumentation.org/packages/zingeR/topics/glmWeightedF'>glmWeightedF</a></code>). For DESeq2, the generalized linear model coefficients are estimated using <code><a href='https://www.rdocumentation.org/packages/DESeq2/topics/nbinomWaldTest'>nbinomWaldTest</a></code> and the weighting is done by setting the degrees of freedom for the T distribution.</p></dd>
<dt>edgeR-ZINB-WaVE, DESeq2-ZINB-WaVE</dt><dd><p>In a first step, a zero-inflated negative binomial regression model  is fitted (see <code><a href='https://www.rdocumentation.org/packages/zinbwave/topics/zinbFit'>zinbFit</a></code>) to estimate observational weights (see <code><a href='https://www.rdocumentation.org/packages/zinbwave/topics/computeObservationalWeights'>computeObservationalWeights</a></code>) used for dispersion estimation in <code><a href='https://www.rdocumentation.org/packages/edgeR/topics/estimateDisp'>estimateDisp</a></code>. For the edgeR approach, the generalized model as implemented in <code><a href='https://www.rdocumentation.org/packages/edgeR/topics/glmfit'>glmFit</a></code> is fitted. This is followed by an adapted LRT for differential testing to account for the weighting (see <code><a href='https://www.rdocumentation.org/packages/zinbwave/topics/glmWeightedF'>glmWeightedF</a></code>). For DESeq2, the generalized linear model coefficients are estimated using <code><a href='https://www.rdocumentation.org/packages/DESeq2/topics/nbinomWaldTest'>nbinomWaldTest</a></code> and the weighting is done by setting the degrees of freedom for the T distribution.</p></dd>
</dl>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='estimateParam.html'>estimateParam</a></code>,  <code><a href='insilicoNBParam.html'>insilicoNBParam</a></code> for negative binomial parameter specifications;<br />
 <code><a href='DESetup.html'>DESetup</a></code>, <code><a href='SimSetup.html'>SimSetup</a></code> for simulation setup;<br />
 <code><a href='evaluateDE.html'>evaluateDE</a></code> for DE evaluation.</p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'># NOT RUN {</span>
<span class='co'>## define DE parameters and set up simulations</span>
<span class='no'>de.opts</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='DESetup.html'>DESetup</a></span>(<span class='kw'>ngenes</span> <span class='kw'>=</span> <span class='fl'>10000</span>, <span class='kw'>nsims</span> <span class='kw'>=</span> <span class='fl'>25</span>,
<span class='kw'>p.DE</span> <span class='kw'>=</span> <span class='fl'>0.2</span>, <span class='kw'>pLFC</span> <span class='kw'>=</span> <span class='kw'>function</span>(<span class='no'>x</span>) <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/sample'>sample</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(-<span class='fl'>1</span>,<span class='fl'>1</span>), <span class='kw'>size</span><span class='kw'>=</span><span class='no'>x</span>,<span class='kw'>replace</span><span class='kw'>=</span><span class='fl'>TRUE</span>)*<span class='fu'><a href='https://www.rdocumentation.org/packages/stats/topics/GammaDist'>rgamma</a></span>(<span class='no'>x</span>, <span class='fl'>3</span>, <span class='fl'>3</span>),
<span class='kw'>p.B</span><span class='kw'>=</span><span class='fl'>0.1</span>, <span class='kw'>bLFC</span> <span class='kw'>=</span> <span class='kw'>function</span>(<span class='no'>x</span>) <span class='fu'><a href='https://www.rdocumentation.org/packages/stats/topics/Normal'>rnorm</a></span>(<span class='no'>x</span>, <span class='kw'>mean</span><span class='kw'>=</span><span class='fl'>0</span>, <span class='kw'>sd</span><span class='kw'>=</span><span class='fl'>1.5</span>), <span class='kw'>bPattern</span><span class='kw'>=</span><span class='st'>"uncorrelated"</span>,
<span class='kw'>sim.seed</span> <span class='kw'>=</span> <span class='fl'>43856</span>)
<span class='no'>sim.opts</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='SimSetup.html'>SimSetup</a></span>(<span class='kw'>desetup</span> <span class='kw'>=</span> <span class='no'>de.opts</span>,
<span class='kw'>params</span> <span class='kw'>=</span> <span class='no'>kolodziejczk_param</span>,
<span class='kw'>spike</span><span class='kw'>=</span><span class='kw'>NULL</span>, <span class='kw'>size.factors</span> <span class='kw'>=</span> <span class='st'>"equal"</span>,
<span class='kw'>downsample</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>geneset</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)
<span class='co'>## run simulations</span>
<span class='no'>sim.res</span> <span class='kw'>&lt;-</span> <span class='fu'>simulateDE</span>(<span class='kw'>n1</span><span class='kw'>=</span><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='fl'>50</span>,<span class='fl'>96</span>,<span class='fl'>384</span>), <span class='kw'>n2</span><span class='kw'>=</span><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='fl'>80</span>,<span class='fl'>96</span>,<span class='fl'>384</span>),
<span class='kw'>sim.settings</span> <span class='kw'>=</span> <span class='no'>sim.opts</span>,
<span class='kw'>DEmethod</span> <span class='kw'>=</span> <span class='st'>'limma-trend'</span>,
<span class='kw'>Normalisation</span> <span class='kw'>=</span> <span class='st'>'scran'</span>,
<span class='kw'>Preclust</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
<span class='kw'>Prefilter</span> <span class='kw'>=</span> <span class='st'>"FreqFilter"</span>,
<span class='kw'>Impute</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
<span class='kw'>spikeIns</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
<span class='kw'>NCores</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
<span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)
<span class='co'># }</span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#prefiltering-prior-to-imputation-normalisation">Prefiltering prior to imputation/normalisation</a></li>

      <li><a href="#imputation-prior-to-normalisation">Imputation prior to normalisation</a></li>

      <li><a href="#normalisation-applied-to-imputed-read-count-matrix">Normalisation applied to (imputed) read count matrix</a></li>

      <li><a href="#differential-testing-using-raw-read-count-matrix">Differential testing using raw read count matrix</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Beate Vieth</p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Beate Vieth.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.2.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

